programs/Makefile

```diff
@@ -92,12 +84,19 @@ ifneq (,$(filter Windows%,$(OS)))
 endif
 	$(CC) $(FLAGS) $^ $(RES_FILE) -o zstd$(EXT) $(LDFLAGS)
 
+zstd-nogz : CPPFLAGS += -DZSTD_LEGACY_SUPPORT=$(ZSTD_LEGACY_SUPPORT)
 zstd-nogz : HAVE_ZLIB=0
 zstd-nogz : zstd-internal
 
-zstd : CPPFLAGS += $(ZLIBCPP)
-zstd : LDFLAGS += $(ZLIBLD)
-zstd : zstd-internal
+zstd : CPPFLAGS += -DZSTD_LEGACY_SUPPORT=$(ZSTD_LEGACY_SUPPORT)
+zstd :
+	@echo "int main(){}" | $(CC) -o have_zlib -x c - -lz && echo zlib found || echo zlib not found
+	@if [ -s have_zlib ]; then \
+         $(RM) have_zlib$(EXT) fileio.o \
+         && $(MAKE) zstd-internal HAVE_ZLIB=1 CFLAGS="$(CFLAGS)" CPPFLAGS="$(CPPFLAGS) -DZSTD_GZDECOMPRESS " LDFLAGS="$(LDFLAGS) -lz" ; \
```

Comment:
I checked all targets and it seems to work fine. 
My solution is only a proposal. If you prefer not to call `make` from `make` I don't see any option besides a global HAVE_ZLIB variable like in your solution.